### PR TITLE
refine tensor doc and add module to_global doc

### PR DESCRIPTION
**Pull request by hjchen2**
- 完善Tensor.to_global接口文档，增加对grad_sbp的描述
截屏2022-03-18 下午5 04 56
截屏2022-03-18 下午5 05 14
截屏2022-03-18 下午5 05 24

- 完善Tensor.to_local接口文档
截屏2022-03-18 下午5 05 35

- 增加Tensor Attributes文档
截屏2022-03-18 下午5 05 56
截屏2022-03-18 下午5 06 07
截屏2022-03-18 下午5 06 18

- 增加Module.to_consistent和Module.to_global接口的文档
截屏2022-03-18 下午5 06 32

### CODE REVIEW

File: python/oneflow/framework/docstr/module.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import oneflow
+from oneflow.framework.docstr.utils import add_docstr
+
+add_docstr(
+    oneflow.nn.Module.to_consistent,
+    """
+    This interface is no longer available, please use .to_global instead
```
Comment (doombeaker) [inline python/oneflow/framework/docstr/module.py:22]:
```suggestion
    This interface is no longer available, please use :func:`oneflow.nn.Module.to_global` instead
```

改成这样子，会有超链接

File: python/oneflow/framework/docstr/module.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    """,
+)
+
+add_docstr(
+    oneflow.nn.Module.to_global,
+    """
+    Convert the parameters to be global.
+    It will apply the same Tensor.to_global operation to each parameter in this module.
```
Comment (doombeaker) [inline python/oneflow/framework/docstr/module.py:30]:
```suggestion
    It will apply the same :func:`oneflow.Tensor.to_global` operation to each parameter in this module.
```

同上

File: python/oneflow/framework/docstr/tensor.py [hunk 3]
```diff
@@ -175,6 +175,13 @@
     """,
 )
 
+add_docstr(
+    oneflow.Tensor.to_consistent,
+    """
+    This interface is no longer available, please use .to_global instead
```
Comment (doombeaker) [inline python/oneflow/framework/docstr/tensor.py:181]:
```suggestion
    This interface is no longer available, please use :func:`oneflow.Tensor.to_global` instead
```

File: python/oneflow/framework/docstr/global_cast.py [hunk 4]
```diff
@@ -21,15 +21,15 @@
     """
     to_global(input, placement=None, sbp=None, grad_sbp=None) -> Tensor
 
-    Cast a local tensor to global tensor or cast a
-    global tensor to another global tensor with
-    different sbp or placement
+    Convert a local tensor to global tensor or convert a global tensor to another global tensor with
```
Comment (chengtbf) [inline python/oneflow/framework/docstr/global_cast.py:24]: 拆成两段

File: python/oneflow/framework/docstr/global_cast.py [hunk 5]
```diff
@@ -21,15 +21,15 @@
...
+    different sbp or placement.
 
 
     Args:
         input (Tensor): the input tensor.
-        placement (flow.placement, optional): the desired placement of returned global tensor. Default: if None, the input tensor must be consistent one and use its own placement.
-        sbp (flow.sbp.sbp or tuple of flow.sbp.sbp, optional): the desired sbp descriptor of returned global tensor. Default: if None, the input tensor must be consistent one and use its own sbp.
+        placement (flow.placement, optional): the desired placement of returned global tensor. Default: if None, the input tensor must be global and use its own placement.
+        sbp (flow.sbp.sbp or tuple of flow.sbp.sbp, optional): the desired sbp descriptor of returned global tensor. Default: if None, the input tensor must be global and use its own sbp. 
+        grad_sbp (flow.sbp.sbp or tuple of flow.sbp.sbp, optional): manually specify the gradient sbp of the operation in the backward pass. Default: if None, the gradient sbp will be infered automatically.
```
Comment (chengtbf) [inline python/oneflow/framework/docstr/global_cast.py:32]: Example 需要至少增加一个例子。 

local_tensor -> global tensor S0 tensor shape 

从 release note 借鉴过来

Comment (chengtbf) [reply python/oneflow/framework/docstr/global_cast.py:32]: S0 的约束；  
B 的约束； Note: B 会发生 rank 0 的数据覆盖

Comment (chengtbf) [reply python/oneflow/framework/docstr/global_cast.py:32]: 使用 api doc 的 NOTE 功能描述上述的重点约束

File: docs/source/tensor.rst [hunk 6]
```diff
@@ -156,7 +156,8 @@ OneFlow Tensor Class
             tan, 
             tanh, 
             tile, 
-            to, 
+            to,
+            to_consistent,
             to_global,
```
Comment (chengtbf) [inline docs/source/tensor.rst:161]: 挪到 Tensor Attributes 页面中

File: python/oneflow/framework/docstr/global_cast.py (hunk 5 above)
Comment (hjchen2) [reply python/oneflow/framework/docstr/global_cast.py]: 更新了，再review一下

File: docs/source/tensor.rst [hunk 7]
```diff
@@ -156,7 +156,8 @@ OneFlow Tensor Class
             tan, 
             tanh, 
             tile, 
-            to, 
+            to,
+            to_consistent,
```
Comment (chengtbf) [inline docs/source/tensor.rst:160]: 这个是按照字典序排的吗，还是随机排的。  to_consistent 是过时接口，是不是放在 to_global 下面比较好。 字典序排的就不用改了

File: python/oneflow/framework/docstr/module.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+add_docstr(
+    oneflow.nn.Module.to_global,
+    """
+    Convert the parameters and buffers to global.
+
+    It performs the same :func:`oneflow.Tensor.to_global` conversion to each parameter and buffer in this module.
+
+
+    Note:
+        This method modifies the module in-place.
```
Comment (chengtbf) [inline python/oneflow/framework/docstr/module.py:35]: 这里要分情况？如果原有的 module 都是 local tensor，这里就不是 inplace 的吧

File: python/oneflow/framework/docstr/module.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    For example:
+
+    .. code-block:: python
+
+        >>> import oneflow as flow
+        >>> m = flow.nn.Conv2d(in_channels=3, out_channels=4, kernel_size=3)
+        >>> m.to_global(placement=flow.placement("cpu", ranks=[0]), sbp=[flow.sbp.split(0)])
+        >>> m.weight.is_global
+        True
```
Comment (chengtbf) [inline python/oneflow/framework/docstr/module.py:52]: 加一个：

```bash
>>> m.bias.is_global
True
```

**Review by chengtbf — APPROVED** [review]

**Comment by chengtbf** [comment]
总体上已经觉得写的很好了~ 值得推广

File: python/oneflow/framework/docstr/module.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import oneflow
+from oneflow.framework.docstr.utils import add_docstr
+
+add_docstr(
+    oneflow.nn.Module.to_consistent,
+    """
+    This interface is no longer available, please use :func:`oneflow.nn.Module.to_global` instead
```
Comment (strint) [inline python/oneflow/framework/docstr/module.py:22]: instead.

少个句号

File: python/oneflow/framework/docstr/module.py (hunk 8 above)
Comment (strint) [reply python/oneflow/framework/docstr/module.py:35]: 是的，local to global无法Inplace

File: python/oneflow/framework/docstr/tensor.py [hunk 11]
```diff
@@ -169,9 +169,121 @@
...
+        - For global tensor
+
+          At least one of placement and sbp is required.
+
+          If placement and sbp are all the same as this tensor's own placement and sbp, then returns this tensor own.
+    
+    Args:
+        placement (flow.placement, optional): the desired placement of returned global tensor. Default: None
+        sbp (flow.sbp.sbp or tuple of flow.sbp.sbp, optional): the desired sbp of returned global tensor. Default: None
+        grad_sbp (flow.sbp.sbp or tuple of flow.sbp.sbp, optional): manually specify the gradient sbp of this operation in the backward pass. If None, the gradient sbp will be infered automatically. It is only used if this tensor is a global tensor. Default: None
```
Comment (strint) [inline python/oneflow/framework/docstr/tensor.py:196]: specify the sbp of this tensor's grad tensor in the backward pass ?

File: python/oneflow/framework/docstr/tensor.py [hunk 12]
```diff
@@ -169,9 +169,121 @@
...
+        >>> input = flow.tensor([0., 1.], dtype=flow.float32, placement=flow.placement("cpu", ranks=[0, 1]), sbp=[flow.sbp.broadcast]) # doctest: +SKIP
+        >>> output = input.to_global(placement=flow.placement("cpu", ranks=[0, 1]), sbp=[flow.sbp.split(0)]) # doctest: +SKIP
+        >>> print(output.size()) # doctest: +SKIP
+        >>> print(output) # doctest: +SKIP
+
+    .. code-block:: python
+
+        >>> # results on rank 0
+        oneflow.Size([2])
+        tensor([0., 1.], placement=oneflow.placement(type="cpu", ranks=[0, 1]), sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.float32)
```
Comment (strint) [inline python/oneflow/framework/docstr/tensor.py:236]: 这里b to s，除了sbp，看起来什么都没变化。

感觉需要体现一下里面的local tensor变化了，也就是自动做了split？

**Review by strint — APPROVED** [review]

File: python/oneflow/framework/docstr/tensor_attributes.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+        oneflow.placement(type="cuda", ranks=[[0, 1], [2, 3]])
+        
+    """
+
+oneflow.sbp.sbp.__doc__ = r"""
+    A ``oneflow.sbp`` is an object representing that how the data of the global tensor is distributed across the ranks of the ``Tensor`` placement.
+
+    ``oneflow.sbp`` includes three types:
+
+        - oneflow.sbp.split(axis)
```
Comment (strint) [inline python/oneflow/framework/docstr/tensor_attributes.py:44]: 发现split接口这里一个问题，torch下，基本都是叫`dim`，tf下才叫`axis`

https://pytorch.org/docs/stable/search.html?q=dim&check_keywords=yes&area=default#

https://stackoverflow.com/questions/62333053/whats-the-difference-between-dim-in-pytorch-and-axis-in-tensorflow

如果在考虑torch的语境的话，这里我们最好改为dim

File: python/oneflow/framework/docstr/tensor_attributes.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+        >>> p = flow.placement(type="cuda", ranks=[[0, 1], [2, 3]])
+        >>> p
+        oneflow.placement(type="cuda", ranks=[[0, 1], [2, 3]])
+        
+    """
+
+oneflow.sbp.sbp.__doc__ = r"""
+    A ``oneflow.sbp`` is an object representing that how the data of the global tensor is distributed across the ranks of the ``Tensor`` placement.
+
+    ``oneflow.sbp`` includes three types:
```
Comment (lixinqi) [inline python/oneflow/framework/docstr/tensor_attributes.py:42]: 这里的表述有点奇怪，oneflow.sbp是一个python module，而oneflow.sbp.sbp是一个类。

``oneflow.sbp.sbp`` is the distribution descritpor class representing  how the data of the global tensor is distributed across the ranks of the ``Tensor`` placement.

There are three types of distribution descriptor instances in module ``oneflow.sbp``:

File: python/oneflow/framework/docstr/module.py (hunk 8 above)
Comment (hjchen2) [reply python/oneflow/framework/docstr/module.py:35]: 这个是说原地修改module，不是改tensor

Comment (clackhan) [reply python/oneflow/framework/docstr/module.py:35]: 这个inplace针对的对象是 module，针对于module是改变自身的操作

File: python/oneflow/framework/docstr/tensor.py (hunk 12 above)
Comment (clackhan) [reply python/oneflow/framework/docstr/tensor.py:236]: 这个在to_local 接口文档中有所体现，这里提及to_local是不是有点越界？

**Comment by clackhan** [comment]
- 增加Tensor Attributes文档(device)
image

image

- 完善Tensor.device接口文档
image

- 完善Tensor.placement接口文档
image

- 完善Tensor.sbp接口文档
image

- 完善Tensor.is_global接口文档
image

- oneflow.to()接口

File: docs/source/tensor.rst (hunk 7 above)
Comment (hjchen2) [reply docs/source/tensor.rst]: 挪到to_local下面去了

File: python/oneflow/framework/docstr/module.py (hunk 10 above)
Comment (hjchen2) [reply python/oneflow/framework/docstr/module.py]: 加上了

File: python/oneflow/framework/docstr/tensor.py (hunk 11 above)
Comment (hjchen2) [reply python/oneflow/framework/docstr/tensor.py]: done
